system/api-auth.go

```diff
@@ -0,0 +1,131 @@
+		return -1, fmt.Sprintf("invalid token %v", apitoken)
+	}
+}
+
+func (application *Application) validateTicketOwnership(authHeader string) (multiSigAddress, authValidationFailureReason string) {
+	timestamp, timestampSignature, ticketHash := extractTicketAuthParams(strings.TrimPrefix(authHeader, "TicketAuth "))
+	if timestamp == "" || timestampSignature == "" || ticketHash == "" {
+		authValidationFailureReason = fmt.Sprintf("invalid ticket auth header value %s", authHeader)
+		return
+	}
+
+	// Ensure that this signature had not been used in a previous authentication attempt.
+	if application.ProcessedTicketChallenges.ContainsChallenge(timestampSignature) {
+		authValidationFailureReason = fmt.Sprintf("disallowed reuse of ticket auth signature %v", timestampSignature)
+		return
+	}
+
+	authTimestamp, err := strconv.Atoi(timestamp)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("invalid ticket auth timestamp value %v", timestamp)
+		return
+	}
+
+	// Ensure that the auth timestamp is not more than
+	// the permitted number of seconds into the past and future.
+	timestampDelta := time.Now().Unix() - int64(authTimestamp)
+	if math.Abs(float64(timestampDelta)) > maxTicketChallengeAge {
+		authValidationFailureReason = fmt.Sprintf("expired or invalid ticket auth timestamp value %v",
+			timestamp)
+		return
+	}
+
+	// confirm that the timestamp signature is a valid base64 string
+	decodedSignature, err := base64.StdEncoding.DecodeString(timestampSignature)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("invalid ticket auth signature %s", timestampSignature)
+		return
+	}
+
+	// Mark this timestamp signature as used to prevent subsequent reuse.
+	application.ProcessedTicketChallenges.AddChallenge(timestampSignature, maxTicketChallengeAge)
+
+	// todo check if ticket belongs to this vsp
+
+	// get user wallet address using ticket hash
+	// todo: may be better to maintain a memory map of tickets-userWalletAddresses
+	ticketInfo, err := application.StakepooldConnMan.GetTicketInfo(ticketHash)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("ticket auth, get ticket info failed: %v", err)
+		return
+	}
+
+	// check if timestamp signature checks out against address
+	addr, err := dcrutil.DecodeAddress(ticketInfo.OwnerFeeAddress)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("ticket auth, unexpected decode address error: %v", err)
+		return
+	}
+
+	valid, err := wallet.VerifyMessage(timestamp, addr, decodedSignature)
+	if err != nil {
+		authValidationFailureReason = fmt.Sprintf("error validating timestamp signature for ticket auth %v", err)
+		return
+	}
+
+	if valid {
+		multiSigAddress = ticketInfo.MultiSigAddress
+	} else {
+		authValidationFailureReason = "invalid timestamp signature, timestamp was not signed using " +
+			"the ticket owner's reward address"
+	}
+	return
+}
+
+func extractTicketAuthParams(authHeader string) (timestampMessage, timestampSignature, ticketHash string) {
```

Comment:
If there are cases where naked returns are a good idea, this would be one. With 3 return variables all of the same type, naming the return variables make this code more readable. Also, looking at the method body, these variables are assigned in a loop, which means the variables must be predeclared. Would it be preferable to declare them in the next line instead?

Reply:
@jholdstock what do you think of the naked return here?

I think it is more readable to name the variables in the return values when there are multiples of the same type.  otherwise you have to keep the order straight in your mind.

I think this function is better naked.

Reply:
Apologies I'm 5 days late to the discussion, but as I mentioned before, there is a distinction between named return params and naked returns. The former is acceptable (preferable imo), the latter should be avoided.